go/spec/spec.go

```diff
@@ -0,0 +1,295 @@
+// describes. The same Database instance is returned every time, unless Close
+// is called. If the Spec is closed, it is re-opened with a new Database.
+func (sp Spec) GetDatabase() datas.Database {
+	if *sp.db == nil {
+		*sp.db = sp.createDatabase()
+	}
+	return *sp.db
+}
+
+// NewChunkStore returns a new ChunkStore instance that this Spec's
+// DatabaseName describes. It's unusual to call this method, GetDatabase is
+// more useful. Unlike GetDatabase, a new ChunkStore instance is returned every
+// time. If there is no ChunkStore, for example remote databases, returns nil.
+func (sp Spec) NewChunkStore() chunks.ChunkStore {
+	switch sp.Protocol {
+	case "http", "https":
+		return nil
+	case "ldb":
+		return getLDBStore(sp.DatabaseName)
+	case "mem":
+		return chunks.NewMemoryStore()
+	}
+	panic("unreachable")
+}
+
+// GetDataset returns the current Dataset instance for this Spec's Database.
+// GetDataset is live, so if Commit is called on this Spec's Database later, a
+// new up-to-date Dataset will returned on the next call to GetDataset.  If
+// this is not a Dataset spec, returns nil.
+func (sp Spec) GetDataset() (ds datas.Dataset) {
+	if sp.DatasetName != "" {
+		ds = sp.GetDatabase().GetDataset(sp.DatasetName)
+	}
+	return
+}
+
+// GetValue returns the Value at this Spec's Path within its Database, or nil
+// if this isn't a Path Spec or if that path isn't found.
+func (sp Spec) GetValue() (val types.Value) {
+	if !sp.Path.IsEmpty() {
+		val = sp.Path.Resolve(sp.GetDatabase())
+	}
+	return
+}
+
+// Href treats the Protocol and DatabaseName as a URL, and returns its href.
+// For example, the spec http://example.com/path::ds returns
+// "http://example.com/path". If the Protocol is not "http" or "http", returns
+// an empty string.
+func (sp Spec) Href() string {
+	proto := sp.Protocol
+	if proto == "http" || proto == "https" {
+		return proto + ":" + sp.DatabaseName
+	}
+	return ""
+}
+
+// Pin returns a Spec in which the dataset component, if any, has been replaced
+// with the hash of the HEAD of that dataset. This "pins" the path to the state
+// of the database at the current moment in time.  Returns itself if the
+// PathSpec is already "pinned".
+func (sp Spec) Pin() (Spec, bool) {
+	var ds datas.Dataset
+
+	if !sp.Path.IsEmpty() {
+		if !sp.Path.Hash.IsEmpty() {
+			// Spec is already pinned.
+			return sp, true
+		}
+
+		ds = sp.GetDatabase().GetDataset(sp.Path.Dataset)
+	} else {
+		ds = sp.GetDataset()
+	}
+
+	commit, ok := ds.MaybeHead()
+	if !ok {
+		return Spec{}, false
+	}
+
+	spec := sp.Protocol + sp.DatabaseName + Separator + "#" + commit.Hash().String()
+	if sp.Path.Path != nil {
+		spec += sp.Path.Path.String()
+	}
+
+	pinned, err := ForPathOpts(spec, sp.Options)
+	d.PanicIfError(err)
+	*pinned.db = *sp.db
+
+	return pinned, true
+}
+
+func (sp Spec) Close() error {
+	db := *sp.db
+	if db == nil {
+		return nil
+	}
+
+	*sp.db = nil
+	return db.Close()
+}
+
+func (sp Spec) createDatabase() datas.Database {
```

Comment:
Only if the protocol isn't recognised, but it shouldn't be possible because `parseDatabaseSpec` will fail (with an error) if it isn't.

However you bring up a good point which is trying to use a Spec which did fail construction, e.g.:
```go
sp, err := ParseDatabaseSpec(...)
sp.GetDatabase()
```
if there actually was an error then `sp` will be `Spec{}` and `GetDatabase()` will `panic("unreachable")`. I do think that panic is the correct behavior here rather than silently returning nil, but I'll test for it.